### PR TITLE
Adding tag grouping support to goas

### PR DESCRIPTION
**Pull request by carmenquan**
This allows us to group headings in our spec and remove information from our description field

### CODE REVIEW

File: parser.go [hunk 1]
```diff
@@ -358,10 +358,11 @@ func (p *parser) parseEntryPoint() error {
 		return fmt.Errorf("can not parse general API information: %v", err)
 	}
 
-	// Security Scopes are defined at a different level in the hierarchy as where they need to end up in the OpenAPI structure,
+	// Security scopes and tag groups are defined at a different level in the hierarchy as where they need to end up in the OpenAPI structure,
 	// so a temporary list is needed.
 	oauthScopes := make(map[string]map[string]string, 0)
-
+	tagGroups := make(map[string][]string)
+	tagGroupKeys := make([]string, 0, len(tagGroups))
```
Comment (jacobScottAllen) [inline parser.go:365]: What is the expectation if a user has a duplicate tag name?

Comment (carmenquan) [reply parser.go:365]: Can you give an example of the scenario?

Comment (jacobScottAllen) [reply parser.go:365]:
```
"x-tagGroups": [
    {
      "name": "Group 1",
      "tags": [
        "Foo",
        "Bar"
      ],
      "x-traitTag": true
    },
    {
      "name": "Group 1",
      "tags": [
        "Vfoo",
        "Foo With Spaces"
      ],
      "x-traitTag": true
    }
  ]
```

I believe this would be an incorrect input, is that right? Just wanted to make sure we could fail gracefully.

Comment (carmenquan) [reply parser.go:365]: yeah good question, this would be incorrect _output_. The input is defined in the examples folder in `main.go`

```
// @Tags "Group 1" "Foo" "description"
// @Tags "Group 1" "Bar" "Baz"
// @Tags "Group 2" "Vfoo" "another description"
// @Tags "Group 2" "Foo With Spaces" "some other description"
```

and the logic on ~L540 of this file checks to see if we've come across a tag group, and if we have, just append the tag to that tag group, and if we haven't, add a new key to the tag group map/key to the array

File: parser.go [hunk 2]
```diff
@@ -524,15 +529,23 @@ func (p *parser) parseEntryPoint() error {
 	return nil
 }
 
-func parseTags(comment string) (*TagDefinition, error) {
+func parseTags(comment string, tagGroups map[string][]string, tagGroupKeys *[]string) (*TagDefinition, error) {
```
Comment (jacobScottAllen) [inline parser.go:538]: This function is getting difficult to understand - can we reformat or comment to make the conditions clear?

File: parser.go [hunk 3]
```diff
@@ -524,15 +529,23 @@ func (p *parser) parseEntryPoint() error {
...
 	re := regexp.MustCompile("\"([^\"]*)\"")
 	matches := re.FindAllStringSubmatch(comment, -1)
-	if len(matches) == 0 || len(matches[0]) == 1 {
-		return nil, fmt.Errorf("Expected: @Tags \"<name>\" [\"<description>\"] Received: %s", comment)
+	if len(matches) != 3 || len(matches[0]) == 1 {
```
Comment (jacobScottAllen) [inline parser.go:541]: Why this change? Why is 3 the number we are checking?

Comment (carmenquan) [reply parser.go:541]: the 3 count comes from "group(1)" "name(2)" "description(3)", if we have less or more then there isn't the right number for parsing and we want to early exit

File: parser.go (hunk 2 above)
Comment (carmenquan) [reply parser.go:538]: Added a comment for the conditions - agreed we can refactor but I think lets take that out of scope for here.

File: parser.go (hunk 3 above)
Comment (jacobScottAllen) [reply parser.go:541]: Can we add a comment with where the 3 count comes from? i.e your "group(1)" "name(2)" "description(3)" would work perfectly imo

Comment (carmenquan) [reply parser.go:541]: I decided to switch over to a method comment with more detail. I want to be careful about over-commenting on code that should be closer to self documenting.

Completely agree this code is due for a refactor or clean up, but not in scope for this change.

**Review by jacobScottAllen — APPROVED** [review]

**Comment by carmenquan** [comment]
Closing this for now, will consider post refactor
